### ISSUE TITLE
Support streaming query results  for large data set

### ISSUE THREAD

**Issue by igdianov** [introproventures/graphql-jpa-query#224]
When querying large data sets, the entity manager will try to load all query results into memory. Then, it will run out of memory in JVM if it can't fit all data. 

Apollo has a [spec ](https://github.com/apollographql/apollo-server/blob/defer-support/docs/source/defer-support.md)for `@defer` directive, allowing declarative control of when individual fields in a single GraphQL query get fulfilled and sent to the client. With `@defer`, the GraphQL execution phase does not wait for deferred fields to resolve, instead returning null as a placeholder, and sending patches to the client as those fields get resolved asynchronously.
 
It should be possible to use JPA's query result stream API to handle large query results via cursors to pull data by chunks from the database and then pass chunks downstream with GraphQL `@defer` directive support to avoid loading the whole result set into memory. 

The library should also prevent to requesting too much data using sensible defaults, but it is not clear what default should be applied to prevent DoS queries, i.e. query fetch graph level, recursive queries?

**Comment by igdianov** [comment]
@molexx Feel free to add your thoughts...

**Comment by molexx** [comment]
@igdianov cool, looks interesting. We have run into a couple of queries which have OOMed the server.

If I understand right the PRed implementation should return exactly the same response content - so we don't need to change anything at the client - but now the max ram used would be per outer row rather than for the entire query response. Nice.

It looks like full `@defer` support would be quite complex as you'd need to build separate queries rather than the one query with multiple joins that is built now. That would be useful for deep nestings.

For simple DoS protection for now how about a configurable limit on the depth of nesting?

And - this is less important now we have streaming - a default and max page size?


I'll see if I can merge this in and give it a go.

**Comment by duiscairt-foobar** [comment]
Hey, we ran into this issue and adapted a fork as follows:

1. Add subscription type to the schema builder
2. Create a streaming Jpa data fetcher which calls the getResultStream() jpa method and then returns a Publisher<ExecutionResult> whose implementation iterates over the stream passing back results.  This way the core graphql-java does the heavy lifting of returning a full entity as specified by the query for each stream result.
3. Implementing a new controller method that returns a StreamingResponse (multipart http 1.1) and processes the Publisher to write results to the output stream as they are requested.

We now return millions of entities via the api without blowing up the heap.

It works fine but has some undesirable side effects.  
1) each row returned is a deeply nested structure - I.e. it’s s complete packet of data one row at a time - so there is alot of duplication of Json data in the envelope structure.  On the plus side each entity can be processed independently of the whole document returned in the standard query.  
2) the schema introspection shows that the api supports subscriptions which it doesn’t really do, though we use the subscription functionality behind the scenes.  Ideally there would be an option to support web sockets for proper subscription support at least as long as the query has data yet to return but http 1.1 chunked response was sufficient for us.

Hope that helps.

**Comment by duiscairt-foobar** [comment]
> @igdianov cool, looks interesting. We have run into a couple of queries which have OOMed the server.
> 
> If I understand right the PRed implementation should return exactly the same response content - so we don't need to change anything at the client - but now the max ram used would be per outer row rather than for the entire query response. Nice.
> 
> It looks like full `@defer` support would be quite complex as you'd need to build separate queries rather than the one query with multiple joins that is built now. That would be useful for deep nestings.
> 
> For simple DoS protection for now how about a configurable limit on the depth of nesting?
> 
> And - this is less important now we have streaming - a default and max page size?
> 
> I'll see if I can merge this in and give it a go.

Won’t this still result in the full result set being held in the controller since the whole result set will be materialised at that point on the heap?

**Comment by igdianov** [comment]
@duiscairt-foobar You are correct. I have tried to use GraphQL `@defer` directive, but it will not help with heap optimization since GraphQL Java will still process the whole result list in memory by iterating the stream before returning it to controller.

There is `@stream` directive extension spec for GraphQL list results. This should be exactly what is needed to support streaming query results with patched responses. but it is not implemented yet: graphql-java/graphql-java#1453

The Subscription approach is the only way to deal with this problem for now until there is `@stream` directive support. 

Give it a spin. Let me know what you think.

**Comment by molexx** [comment]
> Won’t this still result in the full result set being held in the controller since the whole result set will be materialised at that point on the heap?


If the controller processed ResultSet rows individually and wrote each row's GraphQL JSON to the HTTP response's OutputStream as it goes then there would only be one row's worth of data on the heap?

That would be useful because we wouldn't have to change our web client.

**Comment by duiscairt-foobar** [comment]
> > Won’t this still result in the full result set being held in the controller since the whole result set will be materialised at that point on the heap?
> 
> If the controller processed ResultSet rows individually and wrote each row's GraphQL JSON to the HTTP response's OutputStream as it goes then there would only be one row's worth of data on the heap?
> 
> That would be useful because we wouldn't have to change our web client.

@molexx - the issue is that the controller doesn’t get access to the ResultSet stream.  It gets a stream from the graphql-java lib in the form Publisher<ExecutionResult> which is the one processing the ResultSet stream.  I agree it would be useful to allow the client to be agnostic to whether a ResultSet list or stream was used to create the output.  I do also like the SSE approach @igdianov used - though it is a breaking change for the client.

To make it transparent to the client the controller would have to change to return StreamingResponseBody so the output stream can be written to as the graphql publisher<executionresult> stream is read one record at a time.  The following would need to happen.

1. The opening Json structure would need to be written
2. Each record would need to be unwrapped as the graphql-Java lib delivers each record as if it was self contained.  The unwrapped content of each record would have to be written to the outputstream.
3. The Json structure would need to be closed as a final step.

It’s all a bit messy when dealing with Json to do this.  

As a minimum, your client would still need to deal with a chunked response.  That should ideally be transparent depending on your client http library.  However, graphiql (fat client) chokes on this response encoding...

For our project we use the graphql jpa  library but with our own controller which does a version of above.  However our use case is different.  We expose a download endpoint which takes a graphql query giving all the expressive power of that tech  but returns the data as a flattened CSV output instead of Json which is more efficient for the million row+ use case we had.  In our use case it’s ok for the api to be different for mass download vs adhoc queries.

**Comment by igdianov** [comment]
There is a [proposal](https://robrichard.net/jskongress-2019/#43) spec for `@stream` directive that supposed to address streaming list results:

- Similar to @defer, but for lists
- Will load each element of the list as it is ready
 - Show users the first item in the list without waiting for the whole list

**Request**
```
query {
  talks @stream {
    title
    time
    speaker {
      name
    }
  }
}
```
**Initial Response**
```
{
  "data": {
    "talks": []
  }
}
```

**Patch for first talk**
```
{
  "path": ["talks", 0],
  "data":  {
    "title": "Opening Keynote",
    "time": "10:30-11:00",
    "speaker": {
      "name": "Sean Larkin"
    }
  }
}
```

**Patch for second talk**
```
{
  "path": ["talks", 1],
  "data":  {
    "title": "Knock knock, who’s there?",
    "time": "11:30-12:00",
    "speaker": {
      "name": "Sam Bellen"
    }
  }
}
```

**Comment by igdianov** [comment]
It may be possible to use instrumentation to apply `@stream` directive to list results. 

However, streaming large results without flow control shifts the problem to client heap. The streaming should support RSockets to let client signal when it is ready process next batch of data from result stream.

**Comment by igdianov** [comment]
Closed via #229

**Comment by convolutionadan** [comment]
My search led to the answer



https://stackdev.space/questions/pulling-information-from-the-database-using-streaming-technology
